---
title: "Applying MIRA to a Biological Question"
author: "John Lawson"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Applying MIRA to a Biological Question}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Biological Question
Ewing sarcoma is a pediatric cancer characterized by the fusion of the EWS and FLI1 genes which results in a fusion transcription factor. We want to see if Ewing samples can be distinguished from muscle-related samples based on their respective MIRA scores. To accomplish that, we can compare MIRA scores for Ewing and non-Ewing samples using a set of Ewing-specific DNase hypersensitivity sites (DHSs). A set of muscle-specific DHSs will also be used for comparison.
input: single-nucleotide resolution DNA methylation data (from reduced representation bisulfite sequencing) for samples from Ewing sarcoma and a collection of muscle related samples including human skeletal muscle myoblasts, myotubes differentiated from a skeletal muscle myoblasts cell line, primary myoblasts from patients with Facioscapulohumeral Muscular Dystrophy (FSHD), and myotubes from FSHD myoblasts.
input: A set of Ewing-specific DNase hypersensitivity sites, muscle-specific DNase hypersensitivity sites.

## Package workflow

The workflow is the same as the "Methylation-based Inference of Regulatory Activity" vignette but this time we are adding annotation and using multiple samples and region sets:  
1. Start with appropriately formatted single-nucleotide resolution methylation data, region sets -- each with genomic regions corresponding to some biological annotation, and an annotation file that has sample name (sampleName column) and sample type (sampleType column).  
2. Split the regions into bins and aggregate methylation data in those bins to get a MIRA signature for each region set, sample combination.  
3. Calculate MIRA score for each region set, sample combination based on shape of MIRA signature.


```{r, eval = TRUE}
library(MIRA)
library(data.table) #fread, setkey, 
library(GenomicRanges) #GRanges, resize
#data(exampleAnnoDF2)#object name: "exampleAnnoDF2", data.table
```

```{r, eval = FALSE}
caseFiles = c("RRBS_cpgMethylation_EWS_T1.bed", 
              "RRBS_cpgMethylation_EWS_T2.bed", 
              "RRBS_cpgMethylation_EWS_T3.bed")
controlFiles = c("RRBS_cpgMethylation_Hsmmt_1.bed", 
                 "RRBS_cpgMethylation_Hsmmt_2.bed", 
                 "RRBS_cpgMethylation_Hsmmt_3.bed")
RRBSFiles = c(caseFiles, controlFiles)
regionSetFiles = c("sknmc_specific.bed", "muscle-specifichg38.bed")
BSDTList = lapply(X = RRBSFiles, FUN = BSreadBiSeq)
regionSets = lapply(X = regionSetFiles, FUN = fread)
regionSets = lapply(X = regionSets, 
                    FUN = function(x) setNames(x, c("chr", "start", "end")))
#converting list items from data.table to GRanges objects
#strand is not relevant for our region sets
regionSets = lapply(X = regionSets, FUN = 
                        function(x) GRanges(seqnames = x$chr, 
                                              ranges = IRanges(x$start, x$end)))
regionSets[[1]] = resize(regionSets[[1]], 4000, fix = "center")
regionSets[[2]] = resize(regionSets[[2]], 1750, fix = "center")
names(regionSets) <- c("Ewing_Specific", "Muscle_Specific")
```

An important decision is how big to make each region
When it comes to MIRA, lapply is your friend. This allows for distributed computing with mclapply if desired. Next let's divide each region into bins and find methylation that is contained in each bin in each region. Then matching bins will be aggregated over all regions (all bin1's together, all bin2's together, etc.). The function returnMIRABins accomplishes all of this.


```{r, eval = FALSE}
bigBin = lapply(X = BSDTList, FUN = returnMIRABins, GRList = regionSets, 
                binNum = 21, minReads = 0, sampleNameInBSDT = TRUE)
bigBinDT = rbindlist(bigBin)
bigBinDT[, sampleName := gsub("RRBS_cpgMethylation_", "", sampleName)]
```

```{r, eval = FALSE}
#change to eval = TRUE
data(ewingMyobigBinDT) #name: "bigBinDT", has binned data for this example
setkey(exampleAnnoDF2, sampleName)
setkey(bigBinDT, sampleName)
bigBinDT = merge(bigBinDT, exampleAnnoDF2, all.x = TRUE) #adding annotation
plotMIRARegions(binnedRegDT = bigBinDT)
```

Next we calculate MIRA scores based on the MIRA signatures for all samples.

```{r, eval = FALSE}
#change to eval = TRUE when vignette is working
bigBinDT[, methyl := methyl - min(methyl) + .05, by = .(featureID, sampleName)]
plotMIRARegions(binnedRegDT = bigBinDT)
sampleScores = bigBinDT[, .(score = scoreDip(methyl, 
                                             binCount = 21, 
                                             shoulderShift = 5)), 
                        by = .(featureID, sampleName)]
head(sampleScores)
setkey(exampleAnnoDF2, sampleName)
setkey(sampleScores, sampleName)
#adding annotation
sampleScores = merge(sampleScores, exampleAnnoDF2, all.x = TRUE)
plotMIRAScores(sampleScores)
```

## Interpreting the results
We found that the Ewing sarcoma samples had higher MIRA scores for the Ewing-specific DHSs, with the exception of a couple outliers from the muscle group. The results from the muscle-specific DHSs require more interpretation. The results depended a lot on the size of the region surrounding the DHSs and what was counted as the outer edge of the region when calculating the MIRA score (shoulderShift variable).

