---
title: "Applying MIRA to a Biological Question"
author: "John Lawson"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Applying MIRA to a Biological Question}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
(Note: Only the second half of this vignette is evaluated because of the large files involved.)

## Biological Question
Ewing sarcoma is a pediatric cancer characterized by the fusion of the EWS and FLI1 genes which results in a fusion transcription factor. We want to see if Ewing samples can be distinguished from muscle-related samples based on their respective MIRA scores. To accomplish that, we can compare MIRA scores for Ewing and muscle-related samples using a set of Ewing-specific DNase hypersensitivity sites (DHSs). A set of muscle-specific DHSs will also be used for comparison.

### Input: 
* Single-nucleotide resolution DNA methylation data (from reduced representation bisulfite sequencing) for samples from Ewing sarcoma and a collection of muscle related samples including human skeletal muscle myoblasts, myotubes differentiated from a skeletal muscle myoblasts cell line, primary myoblasts from patients with Facioscapulohumeral Muscular Dystrophy (FSHD), and myotubes from FSHD myoblasts. These samples are part of the [ENCODE project](I couldn't find this data on their website: UPDATE). 
* Two sets of regions -- one with Ewing-specific DHSs and a one with muscle-specific DHSs.
All samples and region sets correspond to the hg38 reference genome.

## Package workflow

The workflow is similar to the "Methylation-based Inference of Regulatory Activity" vignette but this time we are adding annotation and using multiple samples and region sets:  
0. Start with appropriately formatted single-nucleotide resolution methylation data, region sets -- each with genomic regions corresponding to some biological annotation, and an annotation file that has sample name (sampleName column) and sample type (sampleType column).
1. Do the following steps for a each region set with a few samples from each condition/sample type and use the results to pick appropriate parameters for the workflow.  
    1.1. Split the regions into bins and aggregate methylation data in those bins to get a MIRA signature for each region set, sample combination.  
    1.2. Calculate MIRA score for each region set, sample combination based on the shape of the MIRA signature.  
2. Based on preliminary results, pick a region size for each region set depending on what the MIRA signatures look like.  
3. Run the full MIRA analysis with chosen region sizes.  
    3.1. Split the regions into bins and aggregate methylation data in those bins to get a MIRA signature for each region set, sample combination.  
    3.2. Calculate MIRA score for each region set, sample combination based on the shape of the MIRA signature.  
    
The below code shows step 3 of the process (the step after choosing region sizes that are appropriate for your data) although the same thing would be done for steps 1 and 2, just with fewer samples and guesses for the region sizes. How to choose these sizes is discussed more below.


```{r, eval = TRUE, message = FALSE}
library(MIRA)
library(data.table) #fread, setkey, 
library(GenomicRanges) #GRanges, resize
data(exampleAnnoDF2)#object name: "exampleAnnoDF2", data.table
```

```{r, eval = FALSE}
# 12 Ewing samples: T1-T12
caseFiles = paste0("RRBS_cpgMethylation_EWS_T", 1:12, ".bed") 
controlFiles = c("RRBS_cpgMethylation_Hsmm_", 
                 "RRBS_cpgMethylation_Hsmmt_", 
                 "RRBS_cpgMethylation_Hsmmfshd_",
                 "RRBS_cpgMethylation_Hsmmtubefshd_")
# 12 muscle related samples, 3 of each type
controlFiles=paste0(controlFiles, rep(1:3, each = 4), ".bed")
RRBSFiles = c(caseFiles, controlFiles)
regionSetFiles = c("sknmc_specific.bed", "muscle-specific.bed")
BSDTList = lapply(X = RRBSFiles, FUN = BSreadBiSeq)
regionSets = lapply(X = regionSetFiles, FUN = fread)
regionSets = lapply(X = regionSets, 
                    FUN = function(x) setNames(x, c("chr", "start", "end")))
#converting list items from data.table to GRanges objects
#strand is not relevant for our region sets
regionSets = lapply(X = regionSets, FUN = 
                        function(x) GRanges(seqnames = x$chr, 
                                              ranges = IRanges(x$start, x$end)))
regionSets[[1]] = resize(regionSets[[1]], 4000, fix = "center")
regionSets[[2]] = resize(regionSets[[2]], 1750, fix = "center")
names(regionSets) <- c("Ewing_Specific", "Muscle_Specific")
```

The resize function from the Genomic Ranges package was used to make the length of the Ewing DHS regions 4000 and the muscle DHS regions 1750. We made the muscle-specific regions smaller because we noticed that the dip in MIRA signatures was pretty narrow for some of the samples and we needed to stretch the dip out (make it wider) in order to accurately measure it with MIRA. Since MIRA scoring counts a few bins in the center of the MIRA signature as the bottom of the dip, the MIRA signature must be wide enough so that the middle 3-4 bins are at the bottom of the dip. If the regions are too small though, there might be more noise since there would be less methylation data for each bin. Also, the regions must be wide enough to capture the outer edges of the dip, where methylation levels stop increasing and level out or decrease, since the outer edges are used for the MIRA score. 
When it comes to MIRA, lapply is your friend (although this is generally true anyway). This allows for distributed computing with mclapply if desired. For the next step, we divide each region into bins and find methylation that is contained in each bin in each region. Then matching bins will be aggregated over all regions (all bin1's together, all bin2's together, etc.). The function returnMIRABins accomplishes all of this. For the sake of symmetry and MIRA scoring, an odd bin number is recommended.


```{r, eval = FALSE}
bigBin = lapply(X = BSDTList, FUN = returnMIRABins, GRList = regionSets, 
                binNum = 21, minReads = 0, sampleNameInBSDT = TRUE)
bigBinDT = rbindlist(bigBin)
bigBinDT[, sampleName := gsub("RRBS_cpgMethylation_", "", sampleName)]
```
(Note: code below is all evaluated)
```{r, eval = TRUE}
data(ewingMyobigBinDT) #name: "bigBinDT", has binned data for this example
# data.table functions are used to add info from annotation object to bigBinDT
setkey(exampleAnnoDF2, sampleName)
setkey(bigBinDT, sampleName)
bigBinDT = merge(bigBinDT, exampleAnnoDF2, all.x = TRUE) #adding annotation
plotMIRARegions(binnedRegDT = bigBinDT)
```

Next we normalize the binned methylation data (the MIRA signature) and calculate MIRA scores for all samples. Normalizing by subtracting the minimum methylation bin value from each sample/region set combination generally brings the center MIRA signatures to a similar methylation value, making the MIRA score more about the shape of the signature and less about the average methylation. 

```{r, eval = TRUE, results = "hide"}
bigBinDT[, methyl := methyl - min(methyl) + .05, by = .(featureID, sampleName)]
```
```{r, eval = TRUE}
plotMIRARegions(binnedRegDT = bigBinDT)
```


```{r, eval = TRUE}
sampleScores = bigBinDT[, .(score = scoreDip(methyl, 
                                             binCount = 21, 
                                             shoulderShift = "auto")), 
                        by = .(featureID, sampleName)]
head(sampleScores)
setkey(exampleAnnoDF2, sampleName)
setkey(sampleScores, sampleName)
#adding annotation
sampleScores = merge(sampleScores, exampleAnnoDF2, all.x = TRUE)
plotMIRAScores(sampleScores)
```

## Interpreting the results
We found that the Ewing sarcoma samples had higher MIRA scores for the Ewing-specific DHSs, with the exception of a couple outliers from the muscle group. The muscle-specific DHSs require more interpretation. The results depended a lot on the size of the regions. With wider regions, the scores were about the same for Ewing and muscle samples (not shown in this vignette). However, the muscle samples did generally have higher MIRA scores with the smaller regions we used. In general, interpretation will depend on the samples and region sets being used.

