---
title: "Methylation-based Inference of Regulatory Activity"
author: "John Lawson"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Methylation-based Inference of Regulatory Activity}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

Note: This vignette gives a broad overview of the MIRA package and some main functions. A more realistic example workflow may be seen in the "Applying MIRA to a Biological Question" vignette.

## MIRA Overview
The Methylation-based Inference of Regulatory Activity package, MIRA for short, aggregates DNA methylation data from regions across the genome that correspond to a regulatory element, giving a summary signature for that element. The shape of this signature and the associated score can be used as a metric to compare regulatory activity in different samples and conditions. MIRA works with genome-scale, single-nucleotide-resolution methylation data such as is generated by reduced representation bisulfite sequencing and whole genome bisulfite sequencing but overcomes the sparsity of methylation that can be found in such data by aggregating across many regions. Since methylation will generally be lower at the location of transcription factor binding compared to surrounding regions (exceptions exist of course), MIRA can be used to infer differential transcription factor activity based on changes in the deepness of the dip in the MIRA signature, which implies increased or decreased binding.

```{r,echo=FALSE}
library(MIRA)
data(exampleBins)
plotMIRARegions(exampleBins)
exScores=exampleBins[,.(score = scoreDip(methyl,binCount=11)),by=.(featureID,sampleName)]
exScores[,sampleType := c("Condition1","Condition2")][] #normally would do join with annoDF
```

## Input Data
### DNA Methylation Data
MIRA takes methylation data after the methylation calling stage and, for a given genomic coordinate (ie the location of the C of a CpG), should have number methylated reads and total number of reads. 
Data types that can be used for this: single nucleotide resolution DNA methylation data that has number of methylated reads and total reads for sites/CpGs such as from reduced representation bisulfite sequencing or whole genome bisulfite sequencing. For compatibility with MIRA, this data should be represented as a data.table for each sample with columns "chr", "start" (the coordinate of the C of the CpG), "hitCount" (number of times methylated), "readCount" (total number of reads covering this site), and "sampleName" (sample identifier).

```{r}
data("GM06990_1_ExampleSet",package="MIRA")
head(exampleBSDT)
```

### Region Sets
A region set used as input for MIRA should be a group of genomic regions that correspond to the same biological annotation, like ChIP peaks for a transcription factor. Many types of region sets may be used with MIRA. This includes ChIP peaks for transcription factors or histone modifications, promoters for a set of related genes, sites of motif matches, and DNase hypersensitivity sites but other types of region sets may work as well. Many such region sets may be found in online repositories and we have pulled together some major sources at http://databio.org/regiondb. For use in MIRA, each region set should be a GRange object and multiple region sets may be passed to MIRA as a GRangesList with each list element being a region set.

```{r,message=FALSE}
data("Gm12878Nrf1_Subset",package="MIRA")
head(exampleRegionSet)
```

## Package Workflow
The general workflow is as follows:  
1. Start with appropriately formatted single-nucleotide resolution methylation data and one or many sets of genomic regions.   
2. Split the regions into bins and aggregate methylation data in those bins to get MIRA signature.  
3. Calculate MIRA score based on shape of MIRA signature.

### The Input
First let's load the example data. exampleBSDT is a data.table containing genomic location, methylation hit number, and read number from bisulfite sequencing of an Epstein Bar Virus-transformed B-lymphocyte cell line. exampleRegionSet is a GRanges object with regions that share a biological annotation, in this case Nrf1 binding in a different EBV-transformed B-lymphocyte cell line.

```{r}
library(MIRA)
data("annoDF",package="MIRA") #name "annoDF", class data.table/data.frame
data("Gm12878Nrf1_Subset",package="MIRA")#name "exampleRegionSet", class GRanges
data("GM06990_1_ExampleSet",package = "MIRA")#name "exampleBSDT", class data.table/dataframe
BSDTList=list(exampleBSDT)
```

### Binning and Aggregation
Next let's divide each region into bins, find methylation that is contained in each bin in each region, and aggregate matching bins over all the regions. The binNum argument determines how many approximately equally sized bins each region is split into. Methylation in the first bin in each region will be aggregated together into one summary first bin, methylation in the second bin in each region will be aggregated together into one summary second bin etc. This will produce a MIRA signature for each set of regions. The MIRA signature will be symmetrical if no strand information is given for the regions (produced by averaging the signature with the reverse of the signature), because the orientation of the regions is arbitrary with respect to biological features (like a promoter for instance) that could be oriented directionally (e.g. 5' to 3'). If strand information is given, regions on the minus strand will be flipped before being aggregated with plus strand regions so the MIRA signature will be in 5' to 3' orientation. The minReads argument in returnMIRABins is used to screen out region sets that have any bins with less than a minimum number of reads. Here the default of 500 is used.

```{r Aggregate_methylation,message=FALSE}
bigBin=lapply(X = BSDTList,FUN = returnMIRABins,GRList=exampleRegionSet,binNum=11,sampleNameInBSDT=TRUE)
bigBinDT=bigBin[[1]]
plotMIRARegions(binnedRegDT = bigBinDT)
```

###Calculating the MIRA Score
To calculate MIRA scores based on the MIRA signatures, we will apply the scoring function, scoreDip, to the data.table containing the methylation aggregated in bins using data.table syntax. scoreDip calculates a score for each group of bins corresponding to a sample and region set. With MIRA's default scoring method--"logratio", scoreDip will first get the average of the methylation of the two outer bins as well as the average of the methylation of the three middle bins. The MIRA score is the log ratio of the outer average divided by the middle average which gives a measure of the deepness of dip in the middle of the MIRA signature. Higher MIRA scores are associated with deeper dips. A flat MIRA signature would have a score of 0.

```{r Scoring}
sampleScores=bigBinDT[,.(score = scoreDip(methyl,binCount=11)),by=.(featureID,sampleName)]
head(sampleScores)
```

###A Note on Annotation
For real uses of MIRA, samples and region sets should be annotated. The data.table for each sample should include a sampleName column while region sets should be given to MIRA in a named list/GRangesList. An annotation file that matches sample name with sample type (sampleType column) can be used to add sample type information to the data.table after aggregating methylation. A sampleType column is not required for the main functions but is used for the plotting functions. See "Applying MIRA to a Biological Question" vignette for a workflow including annotation.

## Interpreting the Results
Regulatory information can be inferred from the MIRA scores and signatures but interpretation depends on the type of region set used. For example, for transcription factor region sets (non-methylation binding TFs),  lower dips in the signatures and higher scores in one condition compared to another suggests that there is more transcription factor binding in the condition with higher scores. The general trend of deeper dips being associated with increased activity may apply to other types of region sets as well.
